### PR TITLE
Use  os.sched_getaffinity to get info about available CPUs

### PR DESCRIPTION
**Pull request by pawelchcki**
when running inside containers with CPU limits python's
multiprocessing.cpu_count() will return total number of CPU's
available on the machine. However in reality the process and its
children will not be able to be scheduled on all CPUs.
This issue is most pronounced with taskset/cpuset is set for the process.
Causing scheduling contention over limited number of CPU's.

In case of CFS cpu limits, its similar, but due to bursty nature, at least
in a short timeframe the processes will be able to be scheduled on all CPUs.
The overall efficiency however will be reduced still.

There is no straightforward way to read the CPU shares assigned via CFS.
So just handling simple case when affinity is explicitly set should be a good start

### CODE REVIEW

File: chia/util/cpus.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import os, multiprocessing
+
+def get_available_cpus():
+    cpu_count = multiprocessing.cpu_count()
+    try:
+        cpu_count = os.sched_getaffinity(0)
+    except AttributeError:
+        pass
+
+    if os.name == "nt" and cpu_count > 61:
```
Comment (hoffmang9) [inline chia/util/cpus.py:10]: Is "nt" generic across all modern Windows versions?

File: chia/wallet/wallet_blockchain.py [hunk 2]
```diff
@@ -95,10 +95,7 @@ async def create(
         self.lock = asyncio.Lock()
         self.coin_store = coin_store
         self.tx_store = tx_store
-        cpu_count = multiprocessing.cpu_count()
-        if cpu_count > 61:
-            cpu_count = 61  # Windows Server 2016 has an issue https://bugs.python.org/issue26903
-        num_workers = max(cpu_count - 2, 1)
```
Comment (Nikolaj-K) [inline chia/wallet/wallet_blockchain.py:101]: You deleted the num_workers here, so the system breaks. Might be that this is the quantity you want to compute in your function.

File: chia/util/cpus.py (hunk 1 above)
Comment (Nikolaj-K) [reply chia/util/cpus.py:10]: @hoffmang9 This exact check is in any case being done when trying to compile the chia blockchain, at the point where the code tries to import this [external file and this line](https://github.com/WoLpH/portalocker/blob/638f67af428413ef6b2a45633b6e4f034b11f272/portalocker/portalocker.py#L9).
I ran into a problem installing the whole thing, as described in https://github.com/Chia-Network/chia-blockchain/pull/5158.

File: chia/util/cpus.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+        cpu_count = 61      # Windows Server 2016 has an issue https://bugs.python.org/issue26903
```
Comment (Nikolaj-K) [inline chia/util/cpus.py:11]: Looks you can spare some magic and alternatively do
```python
def get_available_cpus() -> int:
    try:
        cpu_count = os.sched_getaffinity(0)
    except AttributeError:
        cpu_count = multiprocessing.cpu_count()

    # Note: Windows Server 2016 has an issue https://bugs.python.org/issue26903
    if os.name == "nt":
        MAX = 61
        cpu_count = min(cpu_count, MAX)

    return cpu_count
```

File: chia/wallet/wallet_blockchain.py (hunk 2 above)
Comment (pawelchcki) [reply chia/wallet/wallet_blockchain.py:101]: Thanks! Sorry for this :sweat_smile:  -  I was hoping to run CI before review happens and catches ugly mistakes like these

File: chia/util/cpus.py (hunk 3 above)
Comment (pawelchcki) [reply chia/util/cpus.py]: Good suggestion, I've included it in the change.

**Comment by pawelchcki** [comment]
Thank you both for quick review!

File: chia/util/cpus.py (hunk 1 above)
Comment (hoffmang9) [reply chia/util/cpus.py]: No worries - GH now requires approval for first time contributors to start full CI - hence you got a comment from some serious pain that caused that comment from me in the first place regarding the Python windows issues. My quick review does look like "nt" is safe indeed.

**Comment by mariano54** [comment]
Thanks for your contribution! do you mind fixing this merge conflict?

**Comment by pawelchcki** [comment]
@mariano54 merge conflict resolved :)

**Comment by hoffmang9** [comment]
Closing in favor of https://github.com/Chia-Network/chia-blockchain/pull/17409

**Pull request closed without merging**